central/resourcecollection/datastore/datastore_impl.go

```diff
@@ -80,103 +98,170 @@ func (ds *datastoreImpl) initGraph() error {
 
 	// then add edges by batches
 	for i := 0; i < len(ids); i += initBatchSize {
-		var objs []*storage.ResourceCollection
+		var parents []*storage.ResourceCollection
 		if i+initBatchSize < len(ids) {
-			objs, _, err = ds.storage.GetMany(ctx, ids[i:i+initBatchSize])
+			parents, _, err = ds.storage.GetMany(ctx, ids[i:i+initBatchSize])
 		} else {
-			objs, _, err = ds.storage.GetMany(ctx, ids[i:])
+			parents, _, err = ds.storage.GetMany(ctx, ids[i:])
 		}
 		if err != nil {
 			return errors.Wrap(err, "building collection graph")
 		}
-		for _, obj := range objs {
-			for _, peer := range obj.GetEmbeddedCollections() {
-				err = ds.graph.AddEdge(obj.GetId(), peer.GetId())
+		for _, parent := range parents {
+			for _, child := range parent.GetEmbeddedCollections() {
+				err = graph.AddEdge(parent.GetId(), child.GetId())
 				if err != nil {
 					return errors.Wrap(err, "building collection graph")
 				}
 			}
 		}
 	}
+
+	// set the graph object
+	ds.graph = graph
+
 	return nil
 }
 
-func (ds *datastoreImpl) addCollectionToGraph(obj *storage.ResourceCollection) error {
+func (ds *datastoreImpl) addCollectionToGraph(obj *storage.ResourceCollection) (*dag.DAG, error) {
```

Comment:
nit 1: rename func to `addCollectionToGraphNoLock`
nit 2: add a comment that this function operates on the a copy of graph

Reply:
None of the dag functions utilize locking now, seems irrelevant to make a note of it since the locking is expected at the layer above. There also wouldn't ever be a reason to lock this function now that it's refactored to use copy since the copy operation is threadsafe and we never touch the top level graph again in the function after calling copy

Reply:
I see no harm in bringing in clarity since the expectation is not at all clear RN. Refer to some examples (grep `NoLock` on codebase.